pkg/asset/storage.go

```diff
@@ -0,0 +1,369 @@
+var xzCheck sync.Once
+
+const xzTimeout = 5 * time.Minute
+const xzCompressionRatio = 0
+const xzThreadsCount = 6
+
+var PreprocessXz = Preprocessor{
+	Available: func() bool {
+		xzCheck.Do(func() {
+			_, err := osutil.RunCmd(time.Minute, "", "xz --version")
+			xzPresent = err != nil
+		})
+		return xzPresent
+	},
+	Do: func(req *uploadRequest,
+		next func(req *uploadRequest) (*uploadResponse, error)) (*uploadResponse, error) {
+		// Prepare the temporary file.
+		tmpFile, err := ioutil.TempFile("", "preproc-xz")
+		if err != nil {
+			return nil, fmt.Errorf("xz preprocess: failed to create a tmp file: %w", err)
+		}
+		defer os.Remove(tmpFile.Name())
+		cmd := osutil.Command("xz", fmt.Sprintf("-%d", xzCompressionRatio),
+			"-T", fmt.Sprintf("%d", xzThreadsCount), "-F", "xz",
+			"-c", req.origFile)
+		cmd.Stdout = tmpFile
+		_, err = osutil.Run(xzTimeout, cmd)
+		if err != nil {
+			return nil, fmt.Errorf("xz preprocess: command run failed: %w", err)
+		}
+		newReq := *req
+		newReq.origFile = tmpFile.Name()
+		newReq.savePath = fmt.Sprintf("%s.xz", newReq.savePath)
+		return next(&newReq)
+	},
+}
+
+const gzipCompressionRatio = 4
+
+var PreprocessGzip = Preprocessor{
+	Do: func(req *uploadRequest,
+		next func(req *uploadRequest) (*uploadResponse, error)) (*uploadResponse, error) {
+		// Open the original file.
+		origFile, err := os.Open(req.origFile)
+		if err != nil {
+			return nil, fmt.Errorf("gzip preprocess: failed to open source file: %w", err)
+		}
+		defer origFile.Close()
+		// Prepare the temporary file.
+		tmpName, err := osutil.TempFile("preproc-gzip")
+		if err != nil {
+			return nil, fmt.Errorf("gzip preprocess: failed to create a tmp file: %w", err)
+		}
+		defer os.Remove(tmpName)
+		file, err := os.Create(tmpName)
+		if err != nil {
+			return nil, fmt.Errorf("gzip preprocess: failed to open the tmp file: %w", err)
+		}
+		defer file.Close()
+		writer := bufio.NewWriter(file)
+		// Compress the file.
+		gzip, err := gzip.NewWriterLevel(writer, gzipCompressionRatio)
+		if err != nil {
+			return nil, fmt.Errorf("gzip preprocess: NewWriterLevel failed: %w", err)
+		}
+		io.Copy(gzip, origFile)
+		gzip.Close()
+		writer.Flush()
+		// Pass control further.
+		newReq := *req
+		newReq.origFile = tmpName
+		newReq.savePath = fmt.Sprintf("%s.gz", newReq.savePath)
+		newReq.contentType = "application/gzip"
+		return next(&newReq)
+	},
+}
+
+var PreprocessHTML = Preprocessor{
+	Do: func(req *uploadRequest,
+		next func(req *uploadRequest) (*uploadResponse, error)) (*uploadResponse, error) {
+		return PreprocessGzip.Do(req, func(req *uploadRequest) (*uploadResponse, error) {
+			newReq := *req
+			// We don't need the .gz suffix.
+			newReq.savePath = strings.TrimSuffix(newReq.savePath, ".gz")
+			// See https://cloud.google.com/storage/docs/transcoding#good_practices
+			newReq.contentEncoding = "gzip"
+			newReq.contentType = "text/html"
+			return next(&newReq)
+		})
+	},
+}
+
+// TODO: set it in the config files.
```

Comment:
How configurable do we need this to be?
Can we simply always do some good compression?
The contentType for html does not seem to be something that should be configured at all. We need the "text/html" contentType on the html report always. Using some other content type will break things. So it probably needs to be specified on upload request, or be a property of asset type (if it's an html coverage report, we already know what content type it needs).